integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -668,6 +689,68 @@ public String getFileSystemName() {
     return mFsName;
   }
 
+  @Override
+  public void umount() {
+    // Waiting for in progress async release to finish
+    if (!mReleasingReadEntries.isEmpty()) {
+      LOG.info("Waiting for all in progress file in stream closing to finish");
+      try {
+        CommonUtils.waitFor("async opened files releasing finished", mReleasingReadEntries::isEmpty,
+            WaitForOptions.defaults().setTimeoutMs(MAX_UMOUNT_WAITTIME_MS / 3));
+      } catch (InterruptedException e) {
+        LOG.error("Umount interrupted");
+        Thread.currentThread().interrupt();
+      } catch (TimeoutException e) {
+        LOG.error("Timeout when waiting in progress file in stream closing,"
+            + "the number of closing entries is {}", mReleasingReadEntries.size());
+      }
+    }
+    if (!mReleasingWriteEntries.isEmpty()) {
+      LOG.info("Waiting for all in progress file out stream closing to finish");
+      try {
+        CommonUtils.waitFor("async write releasing finished", mReleasingWriteEntries::isEmpty,
+            WaitForOptions.defaults().setTimeoutMs(MAX_UMOUNT_WAITTIME_MS / 3));
+      } catch (InterruptedException e) {
+        LOG.error("Umount interrupted");
+        Thread.currentThread().interrupt();
+      } catch (TimeoutException e) {
+        LOG.error("Timeout when waiting in progress file out stream closing,"
+            + "the number of closing entries is {}", mReleasingWriteEntries.size());
+      }
+    }
+
+    if (!mOpenFileEntries.isEmpty()) {
+      for (Map.Entry<Long, FileInStream> entry : mOpenFileEntries.entrySet()) {
+        FileInStream is = entry.getValue();
+        try {
+          synchronized (is) {
+            is.close();
+          }
+        } catch (Throwable t) {
+          LOG.error("Failed to close opened file with id {} in umount operation",
+              entry.getKey(), t);
+        }
+      }
+    }
+    // Release operation is async, we need to make sure
+    // all out stream is closed before umount the fuse
+    if (!mCreateFileEntries.isEmpty()) {
```

Comment:
If we already have `mReleasingReadEntries`, why we need to check `mCreateFileEntries` here?

Reply:
you may talk about why we need wait for mCreateFileEntries cleanup here?
The reason is that release() and umount() operation could be called by our users at any given time. 

The mReleasingWriteEntries, take care of the following sequence 
release() started - umount() - release() finished

The mCreateFileEntries wait take care of the following sequence
umount() - release() started and finished.

Same for mReleasingReadEntries and mOpenFileEntries

Reply:
Thanks for the explanation, but I'm not sure when `umount` function will be called before `release` function.
If one process still hold a `fd`. Call `umount` will caused `failed to unmount /mnt/fuse: Device or resource busy`. Refer :https://man7.org/linux/man-pages/man2/umount2.2.html.

So we need to close the fd first then umount the fuse?